HIP Support and additional CMake Refactorings

bjoo opened the pull request:
This PR encapsulates the additions needed for HIP support as well as some additional makefile restructurings

Commit:
Reset maxBlockSize to 256 for now for tip of LLVM-Project

Commit:
First set of merges. Doesn't build

Commit:
Initial merge completed. Not checked for working yet

Commit:
Removed old FFT plans in favor of Kate's approach in generic Kernels (CUDA only just now)

Commit:
Cleanups compile the library but undefined symbol... Need to switch to other branch, so committing state

Commit:
Removed stream create and stream destroy funcitons from API (no longer needed).
Changed cudaEventRecord(X, 0); to cudaEventRecord(X, device::get_default_stream()) in some testing codes

Commit:
Updated Notes

Commit:
Corrected HIP Noted typo in build script

Commit:
Fixed default stream issues in tests

Commit:
Fixed up qudaMemcpyToSymbolAsync to make Multigrid build happy for now. Tests pass with the exception of clover tests. Will investigate this in generic kernels next to see if it is there too or not

Commit:
Resolved conflicts

Commit:
eliminated temporary qudaAPIStream_t

Commit:
UPdated notes

Commit:
HIP Now Builds

Commit:
Set max no of threads per block to 256

Commit:
Code compiles with hip. I have seen blas_test pass with half and single prec.
There is a segfault in blas_test in double prec
dslash_test runs, but the verification falls afoul of a runtime check to do with
block/grid-sizes.

Commit:
Work in progress state

Commit:
Extracted device specific stuff into quda_api_target.h for HIP and CUDA. Then reworked qudaMemcpyKind to be independent and removed it from the target specific headers

Commit:
Fixed an issue in the target CMakeList.txt I introduced when making gauge_fix_ovr.cu target specific

Commit:
qudaDeviceProp_t is apparently no longer needed

Commit:
Added Manually the fix from hotfix/GK_clover_tr_log. Thanks @cpviolator.
CUDA build now passes all of 'make test'

Commit:
Oopsie with rsqrt

Commit:
Initial conflict resolution

Commit:
Another round of merges from @maddyscientist. qudaMemcpyToSymbolAsync is gone

Commit:
Another round of merges from @maddyscientist. qudaMemcpyToSymbolAsync is gone

Commit:
qudaMemcpyToSymbolAsync removed from API

Commit:
Merged new code from feature/generic_kernel

Commit:
Fixed a copy /  cast constructor in complex_quda.h

Commit:
Merge branch 'merge-generic-kernel-hip' into feature/hip-compile-fixes

yaomingamd commented:
Hi Bjoo,
got this link error "../lib/libquda.so: undefined reference to `computeGaugeFixingOVRQuda'" when build tests using rocm3.10 for hip target.

cpviolator committed:
syntactic sugar

cpviolator committed:
compile fixes for tune.cpp

cpviolator committed:
Move alignas ordering

cpviolator committed:
Apply Kate's snippet for dim3 fix, use QUDA error, isolate mma namespace with preprocessor guards

cpviolator committed:
Replace missing rsqrt solution, CUDA->HIP in print strings, remove complex_quda header from multigrid.h and deflation.h, replace a complex integer initialiser. Compiles with clang for HIP using the hip_build.sh

Commit:
HIP Dslash passes. blas_test segfaults with rocm-3.9.0

Commit:
Added an if defined(__HIP_DEVICE_COMPILE__) check to two syncthreads, that
were only guarded by a QUDA_ARCH

cpviolator committed:
CMake tidy up

cpviolator committed:
Merge branch 'feature/hip-compile-fixes' of https://github.com/lattice/quda into feature/hip-compile-fixes

Commit:
Fixed typo

cpviolator committed:
minor clean-up

cpviolator committed:
Merge branch 'feature/hip-compile-fixes' of https://github.com/lattice/quda into feature/hip-compile-fixes

Commit:
Fixed an unintended linebreak in CMakeLists for NV builds

cpviolator committed:
GK merge

cpviolator committed:
remove tune.cpp from HIP target CMake file

cpviolator committed:
remove extraneous qudaEventElapsedTime def, comment out api defs, cast kernel params as int to avoid printf warnings

cpviolator committed:
Add corresponding quda_hip_api file, use native error checking in quda API, guard against MMA references in coarse_op_preconditioned.cu, clean up hip device .cpp. Does not compile with QUDA P2P enaled due to missing symbols for hipIpcOpenEventHandle and hipIpcGetEventHandle

cpviolator committed:
aesthetics

maddyscientist committed:
Fix dumb bug

cpviolator committed:
Merge remote-tracking branch 'origin/feature/generic_kernel' into feature/hip-compile-fixes

cpviolator committed:
Move the canonnical tune_quda.h to include

cpviolator committed:
reinstate gauge_fix_ovr

cpviolator committed:
Hacks to make thrust work with HIP and CUDA, use dim3 in place of make_uint3 for HIP

cpviolator committed:
reinstate gpu_arch, place target definition in parentheses

cpviolator committed:
Consistent target defs

cpviolator committed:
Merge remote-tracking branch 'origin/feature/generic_kernel' into feature/hip-compile-fixes

cpviolator committed:
Changes during the call: .x .y -> real() imag(), hipand error

cpviolator committed:
Delete commented out header

cpviolator committed:
Delete commented out header

cpviolator committed:
Fix name clash in random.cu, remove extra hipranState define, fix -Wdangling-field warning in randon_init.cuh

cpviolator committed:
Remove Ipc function wrappers from hip target

cpviolator committed:
Clean up formatting in quda_api

cpviolator committed:
use ::value_type trait

cpviolator committed:
util_quda.h now even with GK, move a QUDA_TARGET_HIP define to quda_internal.h

cpviolator committed:
SinCos and rsqrt solutions

cpviolator committed:
Remove quda_fp16 header from hip target, remove Double2 from reduce_helper

cpviolator committed:
Restore deleted file

cpviolator committed:
Restore anonymous namespace fix

cpviolator committed:
remove stray comment marker

cpviolator committed:
class -> struct error from last push

cpviolator committed:
Merge GK

Commit:
Removed quda_api_target.h. Fixed milc_interface modified malloc.cpp

Commit:
Merge branch 'feature/hip-compile-fixes' of https://github.com/lattice/quda into feature/hip-compile-fixes

cpviolator committed:
Resore Balint's double2 fix for debugging

Commit:
Code compiles and links

Commit:
Checked whether the Double2 struct hack is still needed. THe answer is yes it is

Commit:
Removing guards around GaugeFixOVR

Commit:
Removed quda_api_target.h

maddyscientist commented on include/kernels/block_orthogonalize.cuh:
```diff
@@ -214,7 +216,7 @@ namespace quda {
 
             nrm = norm_reducer.AllSum(nrm);
-            auto nrm_inv = nrm > 0.0 ? rsqrt(nrm) : 0.0;
+            auto nrm_inv = nrm > 0.0 ? Rsqrt(nrm) : 0.0;
```
Rsqrt -> rsqrt

maddyscientist commented on include/kernels/block_orthogonalize.cuh:
```diff
@@ -1,11 +1,13 @@
+#include <register_traits.h>
 #include <color_spinor_field_order.h>
 #include <block_reduction_kernel.h>
+#include <cub_helper.cuh>
```
not needed

bjoo replied:
Fixed with a6c4f4a7e

maddyscientist commented on include/kernels/block_orthogonalize.cuh:
```diff
@@ -1,11 +1,13 @@
 // this removes ghost accessor reducing the parameter space needed
 #define DISABLE_GHOST true // do not rename this (it is both a template parameter and a macro)
-
+#include <register_traits.h>
```
not needed

maddyscientist commented on include/kernels/block_orthogonalize.cuh:
```diff
@@ -1,11 +1,13 @@
+#include "quda_define.h"
```
not needed

bjoo replied:
Fixed in 243d2ecf7

maddyscientist commented on include/index_helper.cuh:
```diff
@@ -1,4 +1,8 @@
+#include "quda_define.h"
+#if defined (QUDA_TARGET_HIP)
+#include "hip/hip_runtime.h"
+#endif
```
We can remove this once we abstract the blockIdx.x in this file

bjoo replied:
OK. Still seems to contain blockIdx.x.

maddyscientist commented on include/comm_quda.h:
```diff
@@ -31,7 +31,7 @@ extern "C" {
 
   // routines related to direct peer-2-peer access
-  void comm_set_neighbor_ranks(Topology *topo=NULL);
+  void comm_set_neighbor_ranks(Topology *topo=nullptr);
```
Back to `NULL` since this is an extern C function

maddyscientist commented on include/complex_quda.h:
```diff
@@ -473,7 +474,7 @@ template<>
     complex<float>(float2 z)
     : float2(z){}
-
+  
```
inserted space

bjoo replied:
Hopefully fixed w. commit #1092 40eaf3dc6

maddyscientist commented on include/fast_intdiv.h:
```diff
@@ -118,7 +118,7 @@ __host__ __device__ __forceinline__
 {
   int q;
-#ifdef __CUDA_ARCH__
+#if defined( __CUDA_ARCH__ ) 
```
`ifdef __CUDA_ARCH__`

maddyscientist replied:
Please revert this change.

maddyscientist commented on include/kernels/dslash_staggered.cuh:
```diff
@@ -137,6 +137,7 @@ namespace quda
           Vector in = arg.in.Ghost(d, 0, ghost_idx, their_spinor_parity);
           out -= (conj(U) * in);
         } else if (doBulk<kernel_type>() && !ghost) {
+	   // Reported as unused variable
```
delete old comment

bjoo replied:
fixed in fe57af7f8

bjoo commented:
Should we use 0x0? NULL is just a #define in stdlib.h, 0x0 is language agnostic.


--
Balint Joo,  Oak Ridge Leadership Computing Facility,
Oak Ridge National Laboratory
P.O. Box 2008, 1 Bethel Valley Road, Oak Ridge, TN 37831, USA
email: joob AT ornl.gov.  Tel: +1-757-912-0566 (cell, remote)

From: maddyscientist <notifications@github.com>
Reply-To: lattice/quda <reply@reply.github.com>
Date: Wednesday, January 27, 2021 at 7:08 PM
To: lattice/quda <quda@noreply.github.com>
Cc: Balint Joo <bjoo@jlab.org>, Author <author@noreply.github.com>
Subject: [EXTERNAL] Re: [lattice/quda] Feature/hip compile fixes (#1092)
Resent-From: <bjoo@jlab.org>
Resent-Date: Wednesday, January 27, 2021 at 7:08 PM


@maddyscientist commented on this pull request.

________________________________

In include/comm_quda.h<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_lattice_quda_pull_1092-23discussion-5Fr565727880&d=DwMCaQ&c=CJqEzB1piLOyyvZjb8YUQw&r=WG01dcOd8ZFLFFNkrpbj6A&m=o0w7opqYypTH4zsza6NTw1LJVHdL-egYwTns9FZUEL0&s=LNwJuwMJklH5KG5sqizRiTJ8X6gEwdVnVhCF8RcY8mY&e=>:

> @@ -31,7 +31,7 @@ extern "C" {

   Topology *comm_default_topology(void);



   // routines related to direct peer-2-peer access

-  void comm_set_neighbor_ranks(Topology *topo=NULL);

+  void comm_set_neighbor_ranks(Topology *topo=nullptr);

Back to NULL since this is an extern C function

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_lattice_quda_pull_1092-23pullrequestreview-2D577880502&d=DwMCaQ&c=CJqEzB1piLOyyvZjb8YUQw&r=WG01dcOd8ZFLFFNkrpbj6A&m=o0w7opqYypTH4zsza6NTw1LJVHdL-egYwTns9FZUEL0&s=oTOj3f5M0Nm0VOoMxOn-CZ_DA2iAbeFct75V-uWjoW4&e=>, or unsubscribe<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AAEPL2HQ6WXRXCIOGQH2U4TS4CTFVANCNFSM4UV3MD3Q&d=DwMCaQ&c=CJqEzB1piLOyyvZjb8YUQw&r=WG01dcOd8ZFLFFNkrpbj6A&m=o0w7opqYypTH4zsza6NTw1LJVHdL-egYwTns9FZUEL0&s=iUJnPSZtseRrj1jBqH1L84Yp6TtUJ7V4A3aMjw6-ug8&e=>.

weinbe2 commented:
> Should we use 0x0? NULL is just a #define in stdlib.h, 0x0 is language agnostic.

Not a hill I'll die on, but if we couldn't safely assume the existence and value of `NULL` I think far more codes than just QUDA would have a very bad time...

maddyscientist commented:
> > Should we use 0x0? NULL is just a #define in stdlib.h, 0x0 is language agnostic.
> 
> Not a hill I'll die on, but if we couldn't safely assume the existence and value of `NULL` I think far more codes than just QUDA would have a very bad time...

Indeed, please just revert to `NULL`.  The point more being that this is a pointless change, and I want as little of these as possible, e.g., formatting etc.

Commit:
Resolved conflicts. But may not build

Commit:
Removed HEAD version of math_helper.cuh now in include/targets/cuda

Commit:
Merge complete on NV Side. make test passes on NV

Commit:
Hopefully reconciled files in target/hip with target/cuda. Test build next

Commit:
Post SinCos merge compiles

Commit:
Fixed space problem

Commit:
Removed qudaAPIStream helper type. get_cuda_stream\(\) is now internal to the target specific files

Commit:
Removed include of quda_define.h

Commit:
Removed unneeded quda_define.h

Commit:
Removed unneeded cub_helper.h

Commit:
Removed old comment

Commit:
Minor fixes

Commit:
Many PR tidyups. Factored quda_arch.h out of quda_internal.h

bjoo commented on include/device.h:
```diff
@@ -50,14 +52,6 @@ namespace quda
-     */
-    cudaStream_t get_cuda_stream(const qudaStream_t &stream);
-
     /**
```
cudaStream_t is programming model specific. I moved it to target_device.h so it can be internal

bjoo commented on include/fast_intdiv.h:
```diff
@@ -16,6 +16,7 @@
 
 #ifndef _INT_FASTDIV_KJGIUHFG
 #define _INT_FASTDIV_KJGIUHFG
+#include<quda_arch.h>
```
needed for float2 and double2 etc.

maddyscientist replied:
Since this is a file from an external library, it would be better not to edit this directly.  Why wasn't this addition needed before?

bjoo replied:
I believe it got included within another file after quda_internal.h . I added this here because one of the tests which included this file through another file didn't build.

maddyscientist replied:
In that case, we need to reshuffle the headers where it needs to be done.  Alternatively, I could just move this file to the include/target/cuda/externals dir, and you can make an appropriate substitute for HIP.

bjoo replied:
OK. I will check this once I switch to a faster build

bjoo replied:
Reverted this in commit (0aa8f813d) -- still to be pushed

bjoo commented on include/gauge_tools.h:
```diff
@@ -1,5 +1,6 @@
 #include <random_quda.h>
 #include <timer.h>
+#include <quda_arch.h>
```
for vector types and __host__ __device__ etc.

bjoo commented on include/pgauge_monte.h:
```diff
@@ -1,5 +1,6 @@
 #pragma once
 
+#include <quda_arch.h>
```
__device__ __host__ vector types etc.

bjoo commented on include/quda_arch.h:
```diff
@@ -0,0 +1,9 @@
+#include <cuda_runtime.h>
+#elif defined(QUDA_TARGET_HIP)
+#include <hip/hip_runtime.h>
+#endif
```
New file to include runtime headers so they could be moved out of quda_internal.h sadly can't easily hide this in include/targets/ as tests seem to include headers which occasionally needed.

bjoo commented on include/targets/cuda/thrust_target.h:
```diff
@@ -0,0 +1,3 @@
+#pragma once
+#include <thrust/system/cuda/vector.h>
+#include <thrust/system/cuda/execution_policy.h>
```
Thrust is system specific, so I put it into include/targets/

maddyscientist replied:
Please revert this.  Thrust will be expunged from GK by the end of this week, and this will just end up causing conflicts

bjoo replied:
OK, will do.

bjoo replied:
Reverted in 6810f7a28 -- to be pushed

bjoo commented on include/targets/hip/math_helper.cuh:
```diff
@@ -54,7 +54,11 @@ namespace quda {
+#if defined(__HIP_DEVICE_COMPILE__)
       return ::rsqrt(a);
+#else
+      return 1.0/sqrt(a);
```
No host rsqrt for some reason with HIP

bjoo commented on include/targets/hip/thrust_target.h:
```diff
@@ -0,0 +1,3 @@
+#pragma once
```
Target specific thrust header for HIP

bjoo replied:
Removed due to reverting thurst_helper.cuh in 6810f7a28

bjoo commented on include/timer.h:
```diff
@@ -156,21 +156,11 @@ namespace quda {
     QUDA_PROFILE_EVENT_RECORD,       /**< cuda event record  */
     QUDA_PROFILE_EVENT_QUERY,        /**< cuda event querying */
-    QUDA_PROFILE_EVENT_ELAPSED_TIME, /**< cuda profile event elspased time */
     QUDA_PROFILE_STREAM_WAIT_EVENT,  /**< stream waiting for event completion */
```
Removed excess profiling tags since they are not used.

bjoo commented on lib/coarse_op.cuh:
```diff
@@ -176,15 +175,15 @@ namespace quda {
 #endif
         } else {
-	  
+ 
```
Looks mostly like spacing.

bjoo commented on lib/gauge_fix_fft.cu:
```diff
@@ -379,7 +379,7 @@ namespace quda {
     }
   };
-  
+
```
whitespace

bjoo commented on lib/gauge_fix_ovr.cu:
```diff
@@ -75,7 +75,7 @@ namespace quda {
     }                                                                                                                  \
   } else {                                                                                                             \
-    errorQuda("Not implemented for %d", (int)tp.aux.x);			\
+    errorQuda("Not implemented for %d", static_cast<int>(tp.aux.x));			\
```
HIP vector type related static_cast

bjoo commented on lib/multi_blas_quda.cu:
```diff
@@ -143,7 +143,7 @@ namespace quda {
             break;
 #endif
-          default: errorQuda("warp-split factor %d not instantiated", int{tp.aux.x});
+          default: errorQuda("warp-split factor %d not instantiated", static_cast<int>(tp.aux.x));
```
HIP vector type related static cast

bjoo commented on lib/multi_reduce_quda.cu:
```diff
@@ -568,7 +568,7 @@ namespace quda {
           delete[] result_trans;
         } else {
-          errorQuda("Unexpected transpose parameter %d", int(tp.aux.x));
+          errorQuda("Unexpected transpose parameter %d", static_cast<int>(tp.aux.x));
```
HIP vector type based static cast

bjoo commented on lib/targets/hip/malloc.cpp:
```diff
@@ -206,7 +203,7 @@ namespace quda
     track_malloc(DEVICE, a, ptr);
 #ifdef HOST_DEBUG
-    //hipMemset(ptr, 0xff, size);
+    hipMemset(ptr, 0xff, size);
```
uncommented.

bjoo commented on lib/targets/hip/malloc.cpp:
```diff
@@ -245,7 +242,7 @@ namespace quda
     track_malloc(DEVICE_PINNED, a, ptr);
 #ifdef HOST_DEBUG
-    //hipMemset(ptr, 0xff, size);
+    hipMemset(ptr, 0xff, size);
```
Uncommented

bjoo commented on lib/targets/hip/malloc.cpp:
```diff
@@ -289,7 +286,7 @@ namespace quda
     track_malloc(PINNED, a, ptr);
 #ifdef HOST_DEBUG
-    //memset(ptr, 0xff, a.base_size);
+    memset(ptr, 0xff, a.base_size);
```
Uncommented

bjoo commented on lib/targets/hip/malloc.cpp:
```diff
@@ -311,7 +308,7 @@ namespace quda
     track_malloc(MAPPED, a, ptr);
 #ifdef HOST_DEBUG
-   // memset(ptr, 0xff, a.base_size);
+    memset(ptr, 0xff, a.base_size);
```
uncommented

bjoo commented on lib/targets/hip/malloc.cpp:
```diff
@@ -334,7 +331,7 @@ namespace quda
     track_malloc(MANAGED, a, ptr);
 #ifdef HOST_DEBUG
-    //hipMemset(ptr, 0xff, size);
+    hipMemset(ptr, 0xff, size);
```
uncommented

bjoo commented on lib/tune.cpp:
```diff
@@ -765,8 +765,11 @@ namespace quda
             printfQuda("About to call tunable.apply block=(%d,%d,%d) grid=(%d,%d,%d) shared_bytes=%d aux=(%d,%d,%d)\n",
-                       (int)param.block.x, (int)param.block.y, (int)param.block.z, (int)param.grid.x, (int)param.grid.y, (int)param.grid.z,
-                       (int)param.shared_bytes, (int)param.aux.x, (int)param.aux.y, (int)param.aux.z);
+                       static_cast<int>(param.block.x), static_cast<int>(param.block.y),
```
HIP vector type based static casts

bjoo commented on tests/staggered_dslash_test.cpp:
```diff
@@ -283,7 +283,7 @@ DslashTime dslashCUDA(int niter) {
   comm_barrier();
   device_timer.start();
-  
+
```
whitespace

bjoo commented on tests/utils/command_line_params.cpp:
```diff
@@ -82,7 +82,7 @@ double eps_naik = 0.0;
 double clover_coeff = 0.1;
 bool compute_clover = false;
-bool compute_clover_trlog = false;
+bool compute_clover_trlog = true;
```
reset original value post Kate's fix

maddyscientist commented on include/deflation.h:
```diff
@@ -62,7 +66,7 @@ namespace quda {
         if(RV->IsComposite() == false) errorQuda("\nRitz vectors must be contained in a composite field.\n");
      }
-
+    
```
Why are these changes in this file?

bjoo replied:
Looks like a whitespace. I can remove if needed.

bjoo replied:
This looks like a line got duplicated. If this is the only difference I will revert to the GK file

bjoo replied:
reverted in e0e2c71f1 - to be pushed.

maddyscientist commented on include/kernels/gauge_fix_fft.cuh:
```diff
@@ -70,10 +70,10 @@ namespace quda {
   };
 
-  template <typename store_t, QudaReconstructType recon>
+  template <typename Float_, QudaReconstructType recon>
```
Why this change?

bjoo replied:
I thought I fixed this back to store_t

bjoo replied:
I will revert this.

bjoo replied:
Reverted and fixed with 2bc5b001c. - there was one unset result in there that I initialized further down in the file.

maddyscientist commented on include/kernels/unitarize_force.cuh:
```diff
@@ -55,8 +55,8 @@ namespace quda {
-        svd_abs_error(svd_abs_error),
-        threads(u.VolumeCB(), 2, 1) { }
+        svd_abs_error(svd_abs_error), 
+	threads(u.VolumeCB(), 2, 1) { }
```
formatting changed for the worse

bjoo replied:
That is what I get for trying to meet the whitespace needs. It is possible to diff these with spacing ignored apparently. 
Since you will be clang-tidy-ing after anyway, can we do it that way
?

maddyscientist replied:
`git diff` will show these formatting changes

bjoo replied:
Hopefully fixed with b1cb6f61e

maddyscientist commented on tests/gauge_alg_test.cpp:
```diff
@@ -172,7 +173,7 @@ class GaugeAlgTest : public ::testing::Test {
     auto detu = getLinkDeterminant(*U);
     double2 tru = getLinkTrace(*U);
-    printfQuda("Det: %.16e:%.16e\n", detu.x, detu.y);
+    printfQuda("Det: %.16e:%.16e\n", double(detu.x), double(detu.y));
```
use real/imag instead of this cast

bjoo replied:
I think with detu the auto is a double2 and the `tru` is explicitly a double2. However I could rework it so that these initialize a `complex<double>` from which I can use real() and imag(). Fixed in fe2361790

maddyscientist commented on tests/gauge_alg_test.cpp:
```diff
@@ -197,6 +198,7 @@ TEST_F(GaugeAlgTest, Generation)
   if (testgen) { ASSERT_TRUE(CheckDeterminant(detu)); }
 }
 
+#if defined(QUDA_TARGET_CUDA)
```
why is here?

bjoo replied:
Looks like I disabled the landau and coulomb tests for non-CUDA.  It may be ok to revert these now. To avoid casting I  initialize doubles with the .x .y and .z components (plaquette returns a double3). Reenabled with 21ea09d93

maddyscientist commented on lib/targets/cuda/quda_fortran.cpp:
```diff
@@ -0,0 +1,15 @@
+#include <quda.h>
```
I guess this is an oversight and should be deleted?

bjoo replied:
Possibly. I did not touch the fortran stuff at all.

bjoo replied:
Removed this ass discussed in portability meeting

maddyscientist commented on lib/targets/cuda/quda_api.cpp:
```diff
@@ -123,10 +126,10 @@ namespace quda {
       if (search == cache.end()) {
         cache.insert(func);
-        qudaFuncSetAttribute(func, cudaFuncAttributePreferredSharedMemoryCarveout, (int)cudaSharedmemCarveoutMaxShared);
+        cudaFuncSetAttribute(func, cudaFuncAttributePreferredSharedMemoryCarveout, (int)cudaSharedmemCarveoutMaxShared);
```
Why these changes?

bjoo replied:
This is internal to CUDA in a target specific file now. So we can have it either way. I don't think HIP has this carveaout attribute, so it doesn't need to be abstracted

maddyscientist replied:
Indeed, but please revert this change.  I still want this abstracted for (future) profiling reasons.

bjoo replied:
OK. Reverted in e9306c8d8 -- tho it is odd because it is a qudaXXXX() call that is not in quda_api.h and takes cudaFuncAttributes as params...

maddyscientist replied:
Thanks.  Yes I know it might look odd 😉 .  It's to give me an easy handle for profiling the overhead of this function.

maddyscientist commented on lib/milc_interface.cpp:
```diff
@@ -51,7 +51,7 @@ static const int num_colors = sizeof(colors)/sizeof(uint32_t);
 
```
I don't think this file should be changed at all

bjoo replied:
it doesn't look. like it has been.

bjoo replied:
Ah this is just replacing gridDim with commGridDim, no functionality changes.

bjoo replied:
This may end up being a CMake level fix once CMake has full Language Support for HIP

Commit:
Reverted fast_intdiv.h

Commit:
Reverted thrust_helper.cuh and removed thrust_target.h-s

Commit:
Referted deflation.h to GK

Commit:
Reverted typology -- only change was to leave in an initialization

Commit:
Hopefully fixed formating

Commit:
real/imag fix

Commit:
Reenabled coulomb and landau fixing tests

Commit:
Reverted lib/targets/cuda/quda_api.cpp as requested

Commit:
Reverted nullptr to NULL by popular demand

Commit:
Stashing this for now

Commit:
Removed quda_fortran.cpp, as discussed in meeting. Look in lib/interface for the fortran stuff

Commit:
Workaround for clang issue: made plus<T> op() inline

Commit:
Merged

Commit:
Merged most recent. May not work on HIP yet

Commit:
Merge branch 'feature/hip-compile-fixes' of https://github.com/lattice/quda into feature/hip-compile-fixes

Commit:
Attempted to reconcile with NV

Commit:
Merged from generic_kernel

Commit:
MErged

Commit:
Merged

Commit:
Tried setting max threads per block back to 1024.
BLAS Test failed when attempting to use more then 256 threads.
Reset to original

Commit:
Deleted specializations for double2 and double3 plus functor now we have vector_type

Commit:
Redundant target specific tune_quda.h removed

Commit:
Merged feature/generic_kernel NV Stuff. Not tested. Pushing for transfer

Commit:
Churn again, this time with target dispatch. I think I have done all
I can without compiling. More fixes as compile tests reveal them

Commit:
Merge branch 'feature/hip-compile-fixes' of https://github.com/lattice/quda into feature/hip-compile-fixes

Commit:
Merged target-dispatch
 - added warp_size()/2 specialization for HIP targets
 - added default -O1 optimization level for HIP targets
 - in single precision (-DQUDA_PRECISION=4, -DQUDA_MULTIGRID=OFF)
   'fast' build currently all ctest tests pass and blas_test passes

Commit:
cos != sin! D'Uh! All Ctests now pass with both single and double (with -O1)

Commit:
I needed to link against CUDA::coda_driver and Threads to nor fail on link with Green team for some reason

Commit:
Merge branch 'feature/generic_kernel' into gk-merge

Commit:
removed QUDA_ENABLE_P2P Macros

Commit:
Merge branch 'feature/generic_kernel' into gk-merge
Fixed conflicts.
Fixed up kernel.h, target_device.h, kernel_constant_memory.h etc.

Commit:
Moved/uncommented device::get_cuda_stream() to target specific file as part of churn

Commit:
Merged

Commit:
Merged GK branch

 - Removed 256 threads/block maximum
 - Removed warp_size()/2 limitation on gauge_fix_ovr
 - Removed HIP_DYNAMIC_SHARED

churn should be reduced.
All tests (except DWF with Tensor Units) pass (CTest + BLAS test)

Commit:
Merged GK

Commit:
Propagated target-device paradigm to block_reduction_kernel.h and reduction_kernel.h

Commit:
Reworked auto etc

Commit:
Quick push to help Corbin along

Commit:
Suggested fix

Commit:
Fixed typo

Commit:
removed some extra hipGetLastErrors

Commit:
Added a small amount of extra debugging into multigrid.cpp

Commit:
Merge branch 'feature/generic_kernel' into gk-merge

Commit:
Hacked QIO and added diagnostics to the restrictor

Commit:
MG Now seems to pass.

Commit:
Moved power of two trick to block_orthogonalize.cu as well. Better boilerplating available in C++17 with in class initialization

Commit:
Disabled the Tensor core using Dslash Local Policy test for non-QUDA builds -- I was originally going to disable individual tests in it but after that there were none left

Commit:
Added missing device:: to two calls in the QUDA_FAST_COMPILE_REDUCE parts of ifdefs

Commit:
curious compiler issue workaround

Commit:
Set opt back to -O3

Commit:
Uncommented some things that were left commented with FIXMEs

Commit:
CMake mods to find HIP rependencies

Commit:
Resolved conflicts

Commit:
Merged CMakeLists.txt-s

Commit:
Merged feature/generic_kernel

Commit:
Moved GTest Init to first line of main for gauge_alg_test to precede the regular argument processing. This gets rid of the error about not understanding --gtest_output=...

Commit:
Added QUDA_MMA_AVAILABLE macro into quda_arch.h and replaced mixed preprocessor
declarations using it

Commit:
Moved here from include now

Commit:
Resolved conflicts

Commit:
Added Static casts in the std::max function to force both operands to have the same base type

Commit:
Fixed clang issue by changing fabs() to std::fabs()

Commit:
for hipPointerGetProperties, a returned value of hipErrorInvalidValue does not mean failure. It means hip didn't know about the pointer internally, and it fixes up an attribute structure claiming it is a host pointer. I suppose it could in princinple be a device pointer from some other programming model, but that is a more remote likelyhood

Commit:
Merged moderner CMake, including some jitify fixes

Commit:
Merged from feature/moderner-cmake2-mods

Commit:
merged

Commit:
Merge branch 'feature/generic_kernel' into mod-cmake-merge

Commit:
Minor additions to QUDAConfig.cmake.in, and removed some guardings from quda_arch.h

Commit:
Pulled coarse_op_mma_launch.h  in from GK to fix green team build

Commit:
Merged GK

Commit:
Made changed device_comms_pinned_malloc_ on HIP allocate buffers that
are multiples of 2MiB in size following HIP issue:

https://github.com/ROCm-Developer-Tools/HIP/issues/2257

Commit:
Merged GK

Commit:
HIP and/or clang didn't like the possibility of __launch_bounds__(0) so I had to do a bit of mucking with the SFINAE. Actually did this before but it got overwritten in the last GK merge

Commit:
Added mods for QDP-JIT and moved QUDAVersion.cmake to QUDAConfigVersion.cmake as CMake wants

Commit:
Added a special HIP CFlags for block_ortho to shorten compile times

Commit:
HIP Build now compiles with QDP_JIT

Commit:
I did an oopsie with malloc tracking

Commit:
Updated QDP-JIT allocation/deallocation

QDP::gpu_malloc() and QDP::gpu_free() are used to allocate the
entire QDP-JIT Pool rather than individual allocations. I essentially
inlined the calls the hijacking macros call.

There may be ways now to make un-evictable allocations through QDP-JIT
now for buffers and things. I can add those later

Commit:
Merge of GK

Commit:
Hip specifics catchups to GK

Commit:
Fixed a couple of typos

Commit:
Changed unrolling flag option for block_orthogonalize.cu

Commit:
Added device side dp sincos and rsqrt

Commit:
Moved hipIpcGetMemHandle call to be protected by whether we do P2P. This issue arose in use with QDP-JIT. Turning P2P off caused the regular device_malloc to be called, which allocated via QDP-JIT's allocator. This was not compatible with the IPC call. Whne P2P is enabled hipMalloc is used and the situation is safe

Commit:
Moved cudaIpcGetMemHandle call to be protected by whether we do P2P. This issue arose in use with QDP-JIT. Turning P2P off caused the regular device_malloc to be called, which allocated via QDP-JIT's allocator. This was not compatible with the IPC call. When P2P is enabled a driver based malloc  is used and the situation is safe

Commit:
Attempt to set GPU Local rank from Env-vars first.

Works around issue in case gethostname() doesn't work properly.
Separate commit to allow cherry pick

Commit:
Fixed a typo

Commit:
Merge branch 'feature/generic_kernel' into gk-merge

Fixed conflicts in CMakeLists (in favor of our way of linking QMP with QMPConfig.cmake)
Reducer conflict was removing a #pragma once (and commented out quda_arch.h -- may need to put that back. We'll see)
Comm target fix was as discussed with Kate.
lib/targets/cuda/CMakeLists was cleanup.

Changes in other target files I propagated were:
  malloc.cpp (added non-peak statistics and status formatting (MB->MiB))
  quda_api.cpp (removed 2D Memallocs etc)
  propagated comm_target fix to HIP

Rest of the merges were automagic

Commit:
Added back pragma once

Commit:
Commented out reporting of SHMEM based memeory as we dont use it

Commit:
Merged recent GK

Commit:
Removed duplicate option for QUDA_DYNAMIC_CLOVER

Commit:
Merge branch 'feature/generic_kernel' into gk-merge

Commit:
Fix from Kate to deal with dynamic shared allocation atom_t->int

Commit:
Dunno what changed...

Commit:
Removed an hcc_detail related header

Commit:
Merged GK

Commit:
Merged latest GK

Commit:
Followed through on style with the cuda branch: Transformer -> Functor

Commit:
Merge branch 'feature/generic_kernel' into gk-merge

Commit:
Merge branch 'feature/generic_kernel' into gk-merge

Commit:
Fixing this for like the N-th time

Commit:
Merge branch 'gk-merge' into feature/hip-compile-fixes

Catching up again

Commit:
Added launch bounds on most Kernels

Commit:
Resolved conflicts

maddyscientist commented on include/shared_memory_cache_helper.cuh:
```diff
@@ -139,9 +140,15 @@ namespace quda
 
        @param[in] block Block dimensions for the 3-d shared memory object 
     */
+#if defined (QUDA_TARGET_CUDA)
```
Can this now be removed since HIP has constexpr dim3 constructor support now?

bjoo replied:
Need to check

maddyscientist commented on include/kernels/clover_invert.cuh:
```diff
@@ -50,7 +50,12 @@ namespace quda
       using Mat = HMatrix<real, N>;
       double trLogA = 0.0;
 
+#if 0
```
Can this be removed now?

maddyscientist commented on include/thrust_helper.cuh:
```diff
@@ -0,0 +1,54 @@
+#pragma once
```
This file should not be part of QUDA anymore

Commit:
Merge branch 'develop' into merge-develop

Commit:
Added a target_device.h to bring target::dispatch

Commit:
Merged

Commit:
Backed out the horrible while in clover_invert.cuh

Commit:
P2P fix from Mahdieh Ghazimirsaeed at AMD. Thank You!

Commit:
Merged conflicts

Commit:
Added max and min to hip math_helper.cuh

Commit:
Merge branch 'develop' into devel-merge

* Initial Merge: Conflicts resolved.
* HIP Support in CMake now in use
* Not yet completed adding target specific changes into HIP

Commit:
Finished merge, checked on files on lib/targets/hip c.f. their regular counterparts

Commit:
Made QUDA_TARGET_CUDA/HIP in target_xxx.cmake to be PARENT_SCOPE

Commit:
Moved definition of QUDA_TARGET_CUDA/HIP to toplevel makefile

Commit:
Removed cub_helper

Commit:
More cleanup for PR

Commit:
Merge branch 'develop' into merge-develop

Commit:
Added \#include quda_arch.h back into complex_quda.h

Commit:
Merged develop

Commit:
Merge branch 'develop' into merge-develop

Commit:
Added back comment to match develop

Commit:
Added back empty line to match develop

Commit:
Added back empty line to match develop

Commit:
Added back empty line to match develop

Commit:
Added removed empty lines to match develop

Commit:
Format

Commit:
Format

Commit:
Cosmetics, and removing thrust_helper.cuh which is no longer needed

Commit:
Added extra space

Commit:
Added space

Commit:
Fixed formatting

Commit:
Formatting

Commit:
Mostly formatting

Commit:
Copied CUDA specific file over from develop

Commit:
More format and header tidying

Commit:
More format and header tidying

Commit:
More formatting

Commit:
More reconciliation with develop

Commit:
more reconciliation with develop

Commit:
More reconciliation with develop

bjoo commented on tests/CMakeLists.txt:
```diff
@@ -711,14 +711,17 @@ foreach(pol IN LISTS DSLASH_POLICIES)
-                     --Lsdim 4
-                     --gtest_output=xml:dslash_mobius_local{pol2}.xml)
+    # This requires tensor cores so only enable NVIDIA targets
+    if( QUDA_TARGET_CUDA )
```
Is there a version of this test that does not use tensor cores?

mathiaswagner replied:
@hummingtree ?

hummingtree replied:
No there is not a non-tensor-core version of this. Disabling this test is the right thing to do I believe. I wonder what would happen if the user calls this operator with a HIP build?

hummingtree replied:
Ah I see that `QUDA_MMA_AVAILABLE` is only set when the target is CUDA.

bjoo commented on lib/targets/cuda/malloc.cpp:
```diff
@@ -9,8 +9,7 @@
 #ifdef USE_QDPJIT
-#include "qdp_quda.h"
-#include "qdp_config.h"
+#include "qdp_cache.h"
```
New alloc interface in QDP-JIT

bjoo commented on lib/targets/cuda/malloc.cpp:
```diff
@@ -230,10 +229,16 @@ namespace quda
     }
+#else
+    // QDPJIT version -- barfs internally if it fails
+     QDP::QDP_get_global_cache().addDeviceStatic( &ptr , size , true );
```
New cache alloc mechanism in QDP-JIT

Commit:
More reconciliation with develop

Commit:
applied clang-format

Commit:
Removed .prehip files as none of those now remain

Commit:
Made this consistent with the CUDA behaviour. Set the QUDA_GPU_ARCH either via CMake option or env-var

Commit:
Removed porting notes. No longer needed

Commit:
Removed some parent-scope stuff

mathiaswagner commented on QUDAConfig.cmake.in:
```diff
@@ -6,14 +6,31 @@ set(QUDA_QMP @QUDA_QMP@)
+set(QUDA_TARGET_CUDA @QUDA_TARGET_CUDA@)
+set(QUDA_TARGET_HIP  @QUDA_TARGET_HIP@)
+
+set(QUDA_NVSHMEM  @QUDA_NVSHMEM@)
```
Wondering why this is here? nvshmem is consumed by QUDA so no calling app should care about this.

mathiaswagner commented on lib/CMakeLists.txt:
```diff
@@ -150,6 +148,12 @@ if(BUILD_FORTRAN_INTERFACE)
 endif()
 
+
+if(QUDA_TARGET_HIP) 
```
we need a different solution here. I don't like start cluttering the CMakeLists with specific workarounds.

mathiaswagner replied:
Any ideas for that already?

mathiaswagner commented on CMakeLists.txt:
```diff
@@ -125,19 +152,25 @@ endif()
 option(QUDA_DIRAC_WILSON "build Wilson Dirac operators" ${QUDA_DIRAC_DEFAULT})
 option(QUDA_DIRAC_CLOVER "build clover Dirac operators" ${QUDA_DIRAC_DEFAULT})
+# Dynamic inversion saves memory but decreases the flops
+option(QUDA_DYNAMIC_CLOVER "Dynamically invert the clover term for twisted-clover" OFF)
```
this is now QUDA_CLOVER_DYNAMIC

mathiaswagner commented on CMakeLists.txt:
```diff
@@ -148,9 +181,11 @@ option(QUDA_MPI "build the MPI multi-GPU code" OFF)
 
 # OpenBLAS
 option(QUDA_OPENBLAS "enable OpenBLAS" OFF)
+option(QUDA_DOWNLOAD_OPENBLAS "Download OpenBLAS software as requested by QUDA_OPENBLAS" OFF)
```
This has been moved. @bjoo  can you please check that we don't have leftovers from merge conflicts in this file?
Just looking at the diff a few things should not be here.

mathiaswagner commented on CMakeLists.txt:
```diff
@@ -291,15 +333,21 @@ endif()
 set(CMAKE_CXX_STANDARD ${QUDA_CXX_STANDARD})
 
 # define CXX FLAGS
+if(QUDA_TARGET_CUDA)
```
why is this here?

mathiaswagner commented on CMakeLists.txt:
```diff
@@ -69,6 +86,14 @@ if(TARGET_TYPE_VALID LESS 0)
   message(SEND_ERROR "Please specify a valid QUDA_TARGET_TYPE type! Valid target types are:" "${VALID_TARGET_TYPES}")
 endif()
 
+### Dependent option 
```
not sure I like adding this here

mathiaswagner commented on CMakeLists.txt:
```diff
@@ -311,33 +359,34 @@ set(CMAKE_CXX_FLAGS_SANITIZE
 
 enable_language(CXX)
-
+find_package(Threads REQUIRED)
```
seems like an arbitrary place to have it between the CXX and C flags. How about moving that down a bit .

mathiaswagner commented on QUDAConfig.cmake.in:
```diff
@@ -25,12 +42,36 @@ if( QUDA_QIO )
 enable_language(CUDA)
 find_package(CUDAToolkit REQUIRED)
+elseif(QUDA_TARGET_HIP )
+  # HIP Speific CMake
```
no string feelings but wondering how we can avoid cluttering this file as more targets get added

bjoo commented:
Linkage ?
 Best,
B

--
Balint Joo,  Oak Ridge Leadership Computing Facility,
Oak Ridge National Laboratory
P.O. Box 2008, 1 Bethel Valley Road, Oak Ridge, TN 37831, USA
email: joob AT ornl.gov.  Tel: +1-757-912-0566 (cell, remote)

From: Mathias Wagner ***@***.***>
Reply-To: lattice/quda ***@***.***>
Date: Wednesday, March 2, 2022 at 4:21 PM
To: lattice/quda ***@***.***>
Cc: Balint Joo ***@***.***>, Author ***@***.***>
Subject: [EXTERNAL] Re: [lattice/quda] Feature/hip compile fixes (#1092)
Resent-From: ***@***.***>
Resent-Date: Wednesday, March 2, 2022 at 4:21 PM


@mathiaswagner commented on this pull request.

________________________________

In QUDAConfig.cmake.in<hxxps://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_lattice_quda_pull_1092-23discussion-5Fr818111592&d=DwMCaQ&c=CJqEzB1piLOyyvZjb8YUQw&r=WG01dcOd8ZFLFFNkrpbj6A&m=8p5_kifSGSYyu9ONekRjXChkHdBpqpiJgeT8tXPPD4dwPEHYKYo_57ndsQGb23h-&s=vQRI_p3Zja5J0LvuO8QAiLj1-_jC1PGNIoD-Gt453ys&e=>:

> @@ -6,14 +6,31 @@ set(QUDA_QMP @QUDA_QMP@)

 set(QUDA_MPI @QUDA_MPI@)

 set(QUDA_QIO @QUDA_QIO@)

 set(QUDA_OPENMP @QUDA_OPENMP@)

+set(QUDA_QDPJIT @QUDA_QDPJIT@)

+

+set(QUDA_TARGET_CUDA @QUDA_TARGET_CUDA@)

+set(QUDA_TARGET_HIP  @QUDA_TARGET_HIP@)

+

+set(QUDA_NVSHMEM  @QUDA_NVSHMEM@)

Wondering why this is here? nvshmem is consumed by QUDA so no calling app should care about this.

—
Reply to this email directly, view it on GitHub<hxxps://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_lattice_quda_pull_1092-23pullrequestreview-2D898184752&d=DwMCaQ&c=CJqEzB1piLOyyvZjb8YUQw&r=WG01dcOd8ZFLFFNkrpbj6A&m=8p5_kifSGSYyu9ONekRjXChkHdBpqpiJgeT8tXPPD4dwPEHYKYo_57ndsQGb23h-&s=UyZbYMS1EymFGFTY61xjUaLgvN2guK8qESumyyT6AbY&e=>, or unsubscribe<hxxps://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AAEPL2EIRAF4TOLC3NRKXHTU57LT5ANCNFSM4UV3MD3Q&d=DwMCaQ&c=CJqEzB1piLOyyvZjb8YUQw&r=WG01dcOd8ZFLFFNkrpbj6A&m=8p5_kifSGSYyu9ONekRjXChkHdBpqpiJgeT8tXPPD4dwPEHYKYo_57ndsQGb23h-&s=ht1nbpo-TLZH-nzxwzBKAHMTVGQeRm_X2C4gSGXf7H8&e=>.
Triage notifications on the go with GitHub Mobile for iOS<hxxps://urldefense.proofpoint.com/v2/url?u=https-3A__apps.apple.com_app_apple-2Dstore_id1477376905-3Fct-3Dnotification-2Demail-26mt-3D8-26pt-3D524675&d=DwMCaQ&c=CJqEzB1piLOyyvZjb8YUQw&r=WG01dcOd8ZFLFFNkrpbj6A&m=8p5_kifSGSYyu9ONekRjXChkHdBpqpiJgeT8tXPPD4dwPEHYKYo_57ndsQGb23h-&s=gTm-NyFEoSpuqqt5pY1A-t8wc9pTvb19kcZGct0VtJ8&e=> or Android<hxxps://urldefense.proofpoint.com/v2/url?u=https-3A__play.google.com_store_apps_details-3Fid-3Dcom.github.android-26referrer-3Dutm-5Fcampaign-253Dnotification-2Demail-2526utm-5Fmedium-253Demail-2526utm-5Fsource-253Dgithub&d=DwMCaQ&c=CJqEzB1piLOyyvZjb8YUQw&r=WG01dcOd8ZFLFFNkrpbj6A&m=8p5_kifSGSYyu9ONekRjXChkHdBpqpiJgeT8tXPPD4dwPEHYKYo_57ndsQGb23h-&s=BmCnQmYOL3cS8BafTc2clzMu4Zt7ec-N9TXJZNFNbYY&e=>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

bjoo commented:
The test fails, saying there are no tensor cores.


--
Balint Joo,  Oak Ridge Leadership Computing Facility,
Oak Ridge National Laboratory
P.O. Box 2008, 1 Bethel Valley Road, Oak Ridge, TN 37831, USA
email: joob AT ornl.gov.  Tel: +1-757-912-0566 (cell, remote)

From: Jiqun Tu ***@***.***>
Reply-To: lattice/quda ***@***.***>
Date: Wednesday, March 2, 2022 at 4:34 PM
To: lattice/quda ***@***.***>
Cc: Balint Joo ***@***.***>, Author ***@***.***>
Subject: [EXTERNAL] Re: [lattice/quda] Feature/hip compile fixes (#1092)
Resent-From: ***@***.***>
Resent-Date: Wednesday, March 2, 2022 at 4:33 PM


@hummingtree commented on this pull request.

________________________________

In tests/CMakeLists.txt<hxxps://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_lattice_quda_pull_1092-23discussion-5Fr818121573&d=DwMCaQ&c=CJqEzB1piLOyyvZjb8YUQw&r=WG01dcOd8ZFLFFNkrpbj6A&m=kJ-6aeoM0SVEK1VpfOxZiLi_OUiWCDAyCIlWhnYDcp7c_iSNPi6YmhD9DoaeweQH&s=1J2Aw6uH99xozF5ig3N5Wgcim70cdeOxpUNrUbkTEV0&e=>:

> @@ -711,14 +711,17 @@ foreach(pol IN LISTS DSLASH_POLICIES)

         --gtest_output=xml:dslash_mobius_eofa_test_pol${pol2}.xml)



     # MdagM local operator

-    add_test(NAME dslash_mobius_local-policy${pol2}

-             COMMAND ${QUDA_CTEST_LAUNCH} $<TARGET_FILE:dslash_ctest> ${MPIEXEC_POSTFLAGS}

-                     --dslash-type mobius

-                     --test MatPCDagMatPCLocal

-                     --matpc even-even

-                     --dim 2 4 6 8

-                     --Lsdim 4

-                     --gtest_output=xml:dslash_mobius_local{pol2}.xml)

+    # This requires tensor cores so only enable NVIDIA targets

+    if( QUDA_TARGET_CUDA )

No there is not a non-tensor-core version of this. Disabling this test is the right thing to do I believe. I wonder what would happen if the user calls this operator with a HIP build?

—
Reply to this email directly, view it on GitHub<hxxps://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_lattice_quda_pull_1092-23discussion-5Fr818121573&d=DwMCaQ&c=CJqEzB1piLOyyvZjb8YUQw&r=WG01dcOd8ZFLFFNkrpbj6A&m=kJ-6aeoM0SVEK1VpfOxZiLi_OUiWCDAyCIlWhnYDcp7c_iSNPi6YmhD9DoaeweQH&s=1J2Aw6uH99xozF5ig3N5Wgcim70cdeOxpUNrUbkTEV0&e=>, or unsubscribe<hxxps://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AAEPL2GCT4B67IXB2UX43VLU57NDZANCNFSM4UV3MD3Q&d=DwMCaQ&c=CJqEzB1piLOyyvZjb8YUQw&r=WG01dcOd8ZFLFFNkrpbj6A&m=kJ-6aeoM0SVEK1VpfOxZiLi_OUiWCDAyCIlWhnYDcp7c_iSNPi6YmhD9DoaeweQH&s=8e1tPLNeLSx8hA3RUuxpxUAcKY7xygR5bNZwPzSZwnw&e=>.
Triage notifications on the go with GitHub Mobile for iOS<hxxps://urldefense.proofpoint.com/v2/url?u=https-3A__apps.apple.com_app_apple-2Dstore_id1477376905-3Fct-3Dnotification-2Demail-26mt-3D8-26pt-3D524675&d=DwMCaQ&c=CJqEzB1piLOyyvZjb8YUQw&r=WG01dcOd8ZFLFFNkrpbj6A&m=kJ-6aeoM0SVEK1VpfOxZiLi_OUiWCDAyCIlWhnYDcp7c_iSNPi6YmhD9DoaeweQH&s=8DbIKCQL8OAwRL8jXjPmdsqYsqRwIJehZXnTw43TEp4&e=> or Android<hxxps://urldefense.proofpoint.com/v2/url?u=https-3A__play.google.com_store_apps_details-3Fid-3Dcom.github.android-26referrer-3Dutm-5Fcampaign-253Dnotification-2Demail-2526utm-5Fmedium-253Demail-2526utm-5Fsource-253Dgithub&d=DwMCaQ&c=CJqEzB1piLOyyvZjb8YUQw&r=WG01dcOd8ZFLFFNkrpbj6A&m=kJ-6aeoM0SVEK1VpfOxZiLi_OUiWCDAyCIlWhnYDcp7c_iSNPi6YmhD9DoaeweQH&s=ZJs-7S_uxMP6iBX__O_pj7l_e57PBO71EID2TYwX2TY&e=>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

mathiaswagner commented on QUDAConfig.cmake.in:
```diff
@@ -6,14 +6,31 @@ set(QUDA_QMP @QUDA_QMP@)
 set(QUDA_MPI @QUDA_MPI@)
```
Just a thought I want to put down here so I don't forget: Should we move that to the `lib` dir? After all this file describes the usage requirements for libquda . Might also make it easier with the `QUDA_TARGET_XYZ` definitions.

mathiaswagner commented:
And while looking at this PR again: @bjoo I think it is time to update the PR description now that this is almost done.

Commit:
Fixed target specific includes

Commit:
Resolving some comments from Mathias and Jiqun

mathiaswagner commented on QUDAConfig.cmake.in:
```diff
@@ -25,12 +33,18 @@ if( QUDA_QIO )
 
-enable_language(CUDA)
-find_package(CUDAToolkit REQUIRED)
+if( QUDA_QDPJIT )
```
very minor but should we move that above the targets?

mathiaswagner commented on QUDAConfig.cmake.in:
```diff
@@ -25,12 +33,18 @@ if( QUDA_QIO )
 endif()
 
+if( QUDA_TARGET_CUDA )
+  include(${CMAKE_CURRENT_LIST_DIR}/find_cuda_deps.cmake)
```
no strong feelings but wonder if that name is too generic if we put quda specific things in here. But I guess `find_quda_target_<target>_dependencies.cmake` is quite long.

mathiaswagner committed:
Small whitespace correction

Commit:
Moved HIP specific compile flags on block_orthogonalize.cu into target_hip.cmake

Commit:
Renamed find_cuda_deps.cmake and find_hip_deps.cmake

Commit:
Minor typo fix in a comment

mathiaswagner commented on cmake/find_target_cuda_dependencies.cmake:
```diff
@@ -0,0 +1,13 @@
+
+find_dependency(CUDAToolkit REQUIRED)
+
+#if( QUDA_NVSHMEM )
```
I tested this with Chroma and NVSHMEM enabled and it builds and runs fine with these lines commented out.

mathiaswagner reviewed (changes requested):
My only real remaining request is an updated PR description. Otherwise I am happy to approve this. Great work. Thanks a ton @bjoo .

bjoo commented:
Hi @weinbe2 @hummingtree @maddyscientist , I think we have the CMake structure agreed for now. Are there any other requirements prior to merging this PR?

maddyscientist commented on include/targets/hip/load_store.h:
```diff
@@ -0,0 +1,88 @@
+      reinterpret_cast<T *>(ptr)[idx] = value;
+    }
+
+#if 0
```
Why not just delete these lines rather than hide them behind `#if 0`?

bjoo replied:
I think I was keeping these as a placeholder in case I ever found out if hipcc implements them or not. I can remove them.

maddyscientist replied:
These are calls to inline ptx functions that I wrote, so I think it unlikely that hipcc support them 😄 .  Not really needed for CUDA either TBH.

bjoo replied:
ifdef 0-d cpde removed in 6ee7ff

maddyscientist commented on lib/targets/hip/device.cpp:
```diff
@@ -27,136 +23,151 @@ namespace quda
       }
     }
 
+    hipStream_t get_cuda_stream(const qudaStream_t &stream) { return streams[stream.idx]; }
```
You can nuke `get_cuda_stream`.  It was a temporary hack that I've removed.

bjoo replied:
Nuked in a320a8

maddyscientist commented on lib/targets/hip/device.cpp:
```diff
@@ -27,136 +23,151 @@ namespace quda
+  {
+    namespace hip
+    {
+      hipStream_t get_cuda_stream(const qudaStream_t &stream) { return streams[stream.idx]; }
```
nuke `get_cuda_stream`

bjoo replied:
Nuked in. a320a8

maddyscientist commented on cmake/find_target_hip_dependencies.cmake:
```diff
@@ -0,0 +1,20 @@
+# HIP Speific CMake
```
typo

bjoo replied:
fixed in e814dd

maddyscientist reviewed (changes requested):
Great PR @bjoo.  Happy to approve this.  Just a bit of minor cleanup needed (see comments I just left).

Commit:
Removed #ifdef 0-d code

Commit:
Fixed typo in comment

Commit:
Removed commented out code

Commit:
Nuked get_quda_stream

bjoo commented:
I also removed the commented out `if (QUDA_NVSMEM)` library import in find_cuda_dependencies.cmake as is seemingly not needed.

Commit:
Hopefully fixed the get_stream issue

Commit:
Applied clang-format

maddyscientist commented:
@weinbe2 @hummingtree just checking if you folk want to review this before we merge?

weinbe2 commented:
Reviewing now

weinbe2 commented on CMakeLists.txt:
```diff
@@ -125,19 +144,24 @@ endif()
 
 option(QUDA_DIRAC_WILSON "build Wilson Dirac operators" ${QUDA_DIRAC_DEFAULT})
 option(QUDA_DIRAC_CLOVER "build clover Dirac operators" ${QUDA_DIRAC_DEFAULT})
+# Dynamic inversion saves memory but decreases the flops
```
This seems out of place? Also, increases flops, not decreases.

bjoo replied:
Removed in d25026150

bjoo commented:
Yeah that may be leftover from removing DYNAMIC_CLOVER. I can delete the comment.

--
Balint Joo,  Oak Ridge Leadership Computing Facility,
Oak Ridge National Laboratory
P.O. Box 2008, 1 Bethel Valley Road, Oak Ridge, TN 37831, USA
email: joob AT ornl.gov.  Tel: +1-757-912-0566 (cell, remote)

From: Evan Weinberg ***@***.***>
Reply-To: lattice/quda ***@***.***>
Date: Thursday, March 10, 2022 at 10:55 AM
To: lattice/quda ***@***.***>
Cc: Balint Joo ***@***.***>, Mention ***@***.***>
Subject: [EXTERNAL] Re: [lattice/quda] HIP Support and additional CMake Refactorings (#1092)
Resent-From: ***@***.***>
Resent-Date: Thursday, March 10, 2022 at 10:55 AM


@weinbe2 commented on this pull request.

________________________________

In CMakeLists.txt<hxxps://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_lattice_quda_pull_1092-23discussion-5Fr823872741&d=DwMCaQ&c=CJqEzB1piLOyyvZjb8YUQw&r=WG01dcOd8ZFLFFNkrpbj6A&m=Gq2Zr05uPvwxtcuamMrm5wiKfICoyFtwcgf8FAJ_s0FW3RBPr9F5mHlQxJz5gDBz&s=yrfYLawB0dNwVeDB-MDj1n90Ksn992eiCb0_1R-NgQg&e=>:

> @@ -125,19 +144,24 @@ endif()



 option(QUDA_DIRAC_WILSON "build Wilson Dirac operators" ${QUDA_DIRAC_DEFAULT})

 option(QUDA_DIRAC_CLOVER "build clover Dirac operators" ${QUDA_DIRAC_DEFAULT})

+# Dynamic inversion saves memory but decreases the flops

This seems out of place? Also, increases flops, not decreases.

—
Reply to this email directly, view it on GitHub<hxxps://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_lattice_quda_pull_1092-23pullrequestreview-2D906053367&d=DwMCaQ&c=CJqEzB1piLOyyvZjb8YUQw&r=WG01dcOd8ZFLFFNkrpbj6A&m=Gq2Zr05uPvwxtcuamMrm5wiKfICoyFtwcgf8FAJ_s0FW3RBPr9F5mHlQxJz5gDBz&s=OBseCymOQZGHp706ahMQF74HcJZDEWj_EeZcyG-S45k&e=>, or unsubscribe<hxxps://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AAEPL2FRALBF54GOOJCDUN3U7ILP3ANCNFSM4UV3MD3Q&d=DwMCaQ&c=CJqEzB1piLOyyvZjb8YUQw&r=WG01dcOd8ZFLFFNkrpbj6A&m=Gq2Zr05uPvwxtcuamMrm5wiKfICoyFtwcgf8FAJ_s0FW3RBPr9F5mHlQxJz5gDBz&s=PL3Igt34eWUP2njm68QUiJOufpViGeu17Ea9qNpLm5g&e=>.
Triage notifications on the go with GitHub Mobile for iOS<hxxps://urldefense.proofpoint.com/v2/url?u=https-3A__apps.apple.com_app_apple-2Dstore_id1477376905-3Fct-3Dnotification-2Demail-26mt-3D8-26pt-3D524675&d=DwMCaQ&c=CJqEzB1piLOyyvZjb8YUQw&r=WG01dcOd8ZFLFFNkrpbj6A&m=Gq2Zr05uPvwxtcuamMrm5wiKfICoyFtwcgf8FAJ_s0FW3RBPr9F5mHlQxJz5gDBz&s=feOUVzhLsfSW1vJMneur7Pw4lioPv-_Ndv0gMDpd7a0&e=> or Android<hxxps://urldefense.proofpoint.com/v2/url?u=https-3A__play.google.com_store_apps_details-3Fid-3Dcom.github.android-26referrer-3Dutm-5Fcampaign-253Dnotification-2Demail-2526utm-5Fmedium-253Demail-2526utm-5Fsource-253Dgithub&d=DwMCaQ&c=CJqEzB1piLOyyvZjb8YUQw&r=WG01dcOd8ZFLFFNkrpbj6A&m=Gq2Zr05uPvwxtcuamMrm5wiKfICoyFtwcgf8FAJ_s0FW3RBPr9F5mHlQxJz5gDBz&s=Z3JaFo39-YtEsQpqfXxVq0DnLiXhcXVyf0CRLu4MFHI&e=>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

weinbe2 commented on CMakeLists.txt:
```diff
@@ -351,6 +382,7 @@ if(QUDA_CLOVER_RECONSTRUCT AND NOT QUDA_CLOVER_DYNAMIC)
   message(SEND_ERROR "QUDA_CLOVER_RECONSTRUCT requires QUDA_CLOVER_DYNAMIC)")
 endif()
 
+find_package(Threads REQUIRED)
```
What's this package for?

bjoo replied:
-lpthreads

Commit:
Removed comment as per Evan's comment

bjoo commented:
-lpthreads



--
Balint Joo,  Oak Ridge Leadership Computing Facility,
Oak Ridge National Laboratory
P.O. Box 2008, 1 Bethel Valley Road, Oak Ridge, TN 37831, USA
email: joob AT ornl.gov.  Tel: +1-757-912-0566 (cell, remote)

From: Evan Weinberg ***@***.***>
Reply-To: lattice/quda ***@***.***>
Date: Thursday, March 10, 2022 at 11:01 AM
To: lattice/quda ***@***.***>
Cc: Balint Joo ***@***.***>, Mention ***@***.***>
Subject: [EXTERNAL] Re: [lattice/quda] HIP Support and additional CMake Refactorings (#1092)
Resent-From: ***@***.***>
Resent-Date: Thursday, March 10, 2022 at 11:01 AM


@weinbe2 commented on this pull request.

________________________________

In CMakeLists.txt<hxxps://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_lattice_quda_pull_1092-23discussion-5Fr823878401&d=DwMCaQ&c=CJqEzB1piLOyyvZjb8YUQw&r=WG01dcOd8ZFLFFNkrpbj6A&m=ea-Eusvfom62FZYVXj-8_hW-LiZTGjg57fCB9VVWdfkA_p0fZjONo6uY2YwrMCFl&s=AlKYtIRgiTR7exRTqJzOCQgeXbfQArzp989CqWpLv90&e=>:

> @@ -351,6 +382,7 @@ if(QUDA_CLOVER_RECONSTRUCT AND NOT QUDA_CLOVER_DYNAMIC)

   message(SEND_ERROR "QUDA_CLOVER_RECONSTRUCT requires QUDA_CLOVER_DYNAMIC)")

 endif()



+find_package(Threads REQUIRED)

What's this package for?

—
Reply to this email directly, view it on GitHub<hxxps://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_lattice_quda_pull_1092-23pullrequestreview-2D906061849&d=DwMCaQ&c=CJqEzB1piLOyyvZjb8YUQw&r=WG01dcOd8ZFLFFNkrpbj6A&m=ea-Eusvfom62FZYVXj-8_hW-LiZTGjg57fCB9VVWdfkA_p0fZjONo6uY2YwrMCFl&s=KbVSBSpl-hBohhfX--Ppl7bWOraE72sEeBViBFliE-0&e=>, or unsubscribe<hxxps://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AAEPL2CRNUVWKVIKIZIUOQ3U7IMDFANCNFSM4UV3MD3Q&d=DwMCaQ&c=CJqEzB1piLOyyvZjb8YUQw&r=WG01dcOd8ZFLFFNkrpbj6A&m=ea-Eusvfom62FZYVXj-8_hW-LiZTGjg57fCB9VVWdfkA_p0fZjONo6uY2YwrMCFl&s=0p81N1jxiSte0UXQIsNbRpOq5LO_VnDVshbvD5M4ek8&e=>.
Triage notifications on the go with GitHub Mobile for iOS<hxxps://urldefense.proofpoint.com/v2/url?u=https-3A__apps.apple.com_app_apple-2Dstore_id1477376905-3Fct-3Dnotification-2Demail-26mt-3D8-26pt-3D524675&d=DwMCaQ&c=CJqEzB1piLOyyvZjb8YUQw&r=WG01dcOd8ZFLFFNkrpbj6A&m=ea-Eusvfom62FZYVXj-8_hW-LiZTGjg57fCB9VVWdfkA_p0fZjONo6uY2YwrMCFl&s=lg4Ihbm-_c9XW8K7krp8_cmGTu9tHQMKUNfNGStZTZ4&e=> or Android<hxxps://urldefense.proofpoint.com/v2/url?u=https-3A__play.google.com_store_apps_details-3Fid-3Dcom.github.android-26referrer-3Dutm-5Fcampaign-253Dnotification-2Demail-2526utm-5Fmedium-253Demail-2526utm-5Fsource-253Dgithub&d=DwMCaQ&c=CJqEzB1piLOyyvZjb8YUQw&r=WG01dcOd8ZFLFFNkrpbj6A&m=ea-Eusvfom62FZYVXj-8_hW-LiZTGjg57fCB9VVWdfkA_p0fZjONo6uY2YwrMCFl&s=kUnTI0jK0fOImsiF0cEGWyTKWJmTST39BbJCL3w4tNY&e=>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

weinbe2 reviewed (approved):
Did some cursory tests on NVIDIA GPUs and everything looks good, this is good to go on my end!

bjoo commented:
Thanks all for the thorough reviews. and for everyone's hard work with the refactorings. What a journey!!!

cpviolator commented:
@bjoo Awesome stuff!
